### ISSUE TITLE
VSCode: Plugin doesn't work in multi-root workspaces

### ISSUE THREAD

**Issue by deftomat** [apollographql/apollo-tooling#821]
I finally create a working `apollo.config.json` (which is really hard as docs are missing or incorrect).

Config looks like this:
```js
module.exports = {
  client: {
    name: 'dev',
    service: {
      url: 'http://localhost:3130/graphql',
      localSchemaFile: 'packages/api/src/schema.graphql'
    },
    includes: ['packages/web/src/**/*.{ts,tsx,js,jsx,graphql}']
  }
};
```

Everything works as expected until you use [multi-root workspaces](https://code.visualstudio.com/docs/editor/multi-root-workspaces) feature.

In multi-root workspace, plugins throws the following error:
```
Unable to read file packages/api/src/schema.graphql. ENOENT: no such file or directory, open 'packages/api/src/schema.graphql'
```

**Comment by ianks** [comment]
Same here, essentially the apollo plugin is not reading relative to the actual apollo.config.js for me.

**Comment by arolson101** [comment]
I'm getting this, too.
I see in the output with trace bumped up that it seems to be getting the first element in the workspace and using that as the root, so the relative paths get screwed up.  E.g.:

```
[Trace - 2:42:45 PM] Sending request 'initialize - (0)'.
Params: {
    "processId": 74836,
    "rootPath": "c:\\ag\\packages\\app",
    "rootUri": "file:///c%3A/ag/packages/app",
    "capabilities": {
```

I found that if I use paths relative to the 'app' folder then it works:
```js
module.exports = {
  client: {
    service: {
      name: 'ag',
      localSchemaFile: 'dist/schema.json'
    },
    includes: ['**/*.tsx'],
  },
}
```

**Comment by JakeDawkins** [comment]
Is this issue still happening on the latest version of the extension? This has seen a lot of work recently and I assume it's been fixed. I'll close this issue if it's old :)

**Comment by elliottsj** [comment]
I'm still experiencing this issue, using version 1.5.2 of the VSCode extension.

**Comment by deftomat** [comment]
The same for me. It's still not working.

**Comment by ZikDead** [comment]
same, any work around this issue?

**Comment by JakeDawkins** [comment]
I haven't had time to dig into this quite yet, but it's still on my radar

**Comment by intellix** [comment]
Getting this for Angular CLI NX generated projects where you have `apps` and `libs` in root. More info here: https://spectrum.chat/apollo/general/vsc-apollo-only-gives-intellisense-on-files-within-src-directory~0c45a389-9bfb-45ab-8810-81cbe683390f

**Comment by galvesribeiro** [comment]
Hey @JakeDawkins!

As we talked on Spectrum, yeah, I'm on latest version and on OSX and the service schema doesn't work either.

**Comment by JakeDawkins** [comment]
So apparently I was mistaking a "monorepo" as a multi-root workspace 🙈. We haven't built proper support for them quite yet into the extension

**Comment by aleccool213** [comment]
@JakeDawkins Are there plans to do this? Where would someone look to start working on this?

**Comment by timdeng2324** [comment]
any idea when this might be released?

**Comment by JakeDawkins** [comment]
@timdeng2324 Hi! 

This is something that we definitely want to get supported, and I agree it'd be a great experience! As we released federation, there were some questions that got raised internally about how multi-root projects would interact with each other in a more dynamic way that also requires some config and project setup changes that we're not quite ready to rebuild. 

**tldr**; we _will_ have this at some point, but it's not a feature with a set timeframe or schedule to work on it.

For the above reasons, I'd also advise anyone else against trying to implement this on your own. It's a spiderweb of dependency questions 😬😬

**Comment by iki** [comment]
@JakeDawkins do we really need anything complex?

My primary goal for fixing the relative `localSchemaFile` by resolving the path as relative to `apollo.config.js` reported in #1666 and https://github.com/apollographql/apollo-tooling/issues/669#issuecomment-484400735 was to avoid the popping extension errors in VS Code, when working in a multi-root workspace, where one of the project parts uses `apollo` CLI. This alone would be a progress and would allow people to use apollo CLI as documented without temporary workaround hacks like https://github.com/apollographql/apollo-tooling/issues/669#issuecomment-551451092. Importantly, resolving the import relatively is a correct way and it doesn't break anything.

Secondary goal would be to make the extension work exactly as `apollo` CLI in multi-root workspace, ie. allow the operations for each of the subprojects based on its `apollo.config.js`. It doesn't need to be handled in the same issue thought, and can be prioritised separately.

Anything more complex? Maybe later, if there're good use-cases. In my case, the multi-root env consists of react-apollo frontend, apollo-server backend, .net backend and docker-compose env, so I don't really need anything above the previous goals. My 2c.

**Comment by tsaiDavid** [comment]
Still experiencing this in our Lerna / Yarn Workspaces monorepos, with various client applications all with their individual `apollo.config.js`

**Comment by iamchathu** [comment]
Our project is structured as Lerna mono repo with multiple CRA projects which interacts with different graphQL endpoints. I have setup in `apollo.config.js` with 

```
module.exports = {
  client: {
    service: {
      name: 'project-1-graphql',
      url: 'http://localhost:4001/graphql',
    },
  },
};

```

But haven't getting any intellisense for it.

**Comment by th-ko** [comment]
@JakeDawkins  Any news on this one? Maybe this would help:
https://github.com/microsoft/vscode/wiki/Adopting-Multi-Root-Workspace-APIs#do-i-need-to-do-anything

**Comment by danielo515** [comment]
You don't need a monorepo to experience this issue. A normal "client/server" folder structure is enough to make the extension fail.
The solution will be to read the paths relative to the apollo.config.js file, and that will solve 99% of the problems. The apollo cli already does this.

**Comment by digeomel** [comment]
I think I have found a temporary workaround. Let's say your have two projects under your root folder, so it's like `root/projects/projA` and `root/projects/projB`. The `schema.graphql` and the `apollo.config.js` files are already there under `root` (and of course it doesn't work in the sub-projects).

The solution is to copy both files under each sub-project, and then I add those folders to the vscode workspace (in addition to the existing root one). When you add the individual folders in the workspace, vscode seems to treat them as separate projects, and it works without any noticeable side-effects. Actually I didn't try the 2nd sub-project, but if it doesn't work, you can just create different workspaces for each sub-project.

Apollo complains about duplicate schema definitions, but in the end works fine. It's a hack, but it seems to work, for me at least.

**Comment by benasher44** [comment]
Is there documentation somewhere on how to get it working with a non-multi-root-but-still-client-folder-server-folder setup?

**Comment by phifa** [comment]
@JakeDawkins  anything we can do to stop these errors?

**Comment by masterkain** [comment]
just finished setting up a Multi-root Workspace for my project and yeah, one of these project folders is a vue application with an `apollo.config.js` that the extension fails to recognize.

I guess I have to go back opening the project folder directly as root

**Comment by pirtlj** [comment]
It is shocking that this has not been fixed yet,  this is a great tool but not supporting multi root workspaces -- which is very common for modern dev orgs -- is really painful.

**Comment by trevor-scheer** [comment]
The vscode extension no longer lives in this repo. Please reopen and reference this issue https://github.com/apollographql/vscode-graphql

**Comment by intellix** [comment]
This now works since I don't know when:
```
module.exports = {
  client: {
    service: 'my-service@production',
    includes: ['{apps,libs}/**/*.{ts,tsx,js,jsx,graphql,gql}']
  },
};
```

**Comment by AaronMBMorse** [comment]
@intellix It appears to be working (not crashing) but I don't have autocomplete. Is autocomplete working for you?

**Comment by intellix** [comment]
yeah that's exactly what I witnessed. I even took a screenshot of it working and celebrated it to the team

**Comment by yoavbls** [comment]
Until `localSchemaFile` is relative to the file you can use this workaround:
 ```ts
module.exports = {
  client: {
    service: {
      name: 'schemafile',
      localSchemaFile: `${__dirname}/schema.graphql`
    }
  }
}
```
